### PR TITLE
WIP PR

### PR DESCRIPTION
**Pull request by itsbaska**
Just creating a PR for you to review my current progress.

### CODE REVIEW

File: src/main/java/server/Server.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      System.out.println("Closing connection with testClient: " + clientSocket.getInetAddress());
+    }
+  }
+
+  public void handleRequest(BufferedReader in, PrintWriter out) {
+    try {
+      Response response = new Response();
+      String line;
+      ArrayList<String> responseText = new ArrayList();
+      while ((line = in.readLine()) != null) {
```
Comment (esmevane) [inline src/main/java/server/Server.java]: I see this is a `BufferedStreamReader`.  [Check out these docs](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html).

When you use `readLine()`, you're imagining the request body as a series of lines of _text_.  _Usually_ that's the case, but it makes some assumptions that it will _always_ be the case.  

Check out these two methods, as well:
* [BufferedStreamReader#ready](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#ready--)
* [BufferedStreamReader#read](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#read--)

The `ready` method tells you whether or not there's anything left to read.  In other words, it does the `!= null` check for you.

And `read` gives you the next character (as an `int` value, so you can coerce it yourself) of the stream.  Basically, what this method is doing, is letting you have access to the buffer of input without making assumptions about its shape.

Take this method for example:

```java
public String readRequestFromSocket(Socket clientSocket) {
  StringBuilder requestBuilder = new StringBuilder();
  BufferedStreamReader buffer = new BufferedStreamReader(
    new InputStreamReader(clientSocket.getInputStream(), StandardCharsets.UTF_8)
  );

  while (buffer.ready() || requestBuilder.length() == 0) {
    requestBuilder.append((char) buffer.read());
  }

  return requestBuilder.toString();
}
```

This does the following:
* Forces an initial attempt to read from a buffer if we haven't yet started our string builder (`requestBuilder.length() == 0`).
* Extract a string from a buffer, character by character (making no assumptions about request body shape).
* Coerces `int` stream into `char` values.  If we wanted to make a different marshaling system here, we could.  Meanwhile, we're also reading those `int` values out as `UTF_8`, which is a handy way to reduce encoding problems.

But even if you don't do it _exactly_ like this, it also does one other important thing: it splits your handling of input reading apart from output reading.

Basically, the writing and the reading are living together in the same loop here, and that's definitely going to cause problems.  Splitting them apart is probably a good first move for you:

1. First, get all of your request.
2. Interpret the request - my favorite thing to do here is to create some of my own code which reads the string content of the request, and then turns what I learned into plain Java objects.
3. Use the interpreted request to create a response.
4. Write the response back to the socket.

File: src/main/java/server/Server.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            if (line.toLowerCase().contains("hello".toLowerCase())) {
+              out.write((response.res200("hello")));
+            } else if (line.toLowerCase().contains("POST".toLowerCase())) {
+              out.write((response.res200("goodbye")));
+            }
+            out.flush();
+        }
+      }
+    } catch (IOException ex) {
+      System.out.println(ex);
```
Comment (esmevane) [inline src/main/java/server/Server.java]: This is a pretty common pattern - maybe there's an opportunity for a `ExitWithMessage` Java object?  So you can do something like `(new ExitWithMessage(ex, message)).perform()`?  Then if you do a different exit strategy later you can change it only once.

There's also a standard recommendation in Java land to _not_ rely on `try` / `catch` structures.  I don't personally mind them, but there are some good reasons to avoid them until absolutely necessary.  The `Exception` objects you get are generally useful in their own way.

One good thing to read here is [Exceptional Ruby](http://exceptionalruby.com/).  While it's a Ruby book, a lot of the good object practices from Ruby come from Java.

File: src/main/java/server/Server.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      System.out.println("Accept failed: " + PORT);
+      System.exit(-1);
+    }
+    return socket;
+  }
+
+  public ServerSocket createSocket() {
+    ServerSocket serverSocket = null;
+    try {
+      serverSocket = new ServerSocket(PORT);
```
Comment (esmevane) [inline src/main/java/server/Server.java]: A lot of these create methods _feel_ like they could be in a single method.  Is the try/catch guarding related to the handling of different exceptions which come from these objects?

File: src/main/java/server/Server.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    return new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
+  }
+
+  public PrintWriter createWriter(Socket clientSocket) throws IOException {
+    return new PrintWriter(new OutputStreamWriter(clientSocket.getOutputStream()));
+  }
+
+  public void closeConnection(BufferedReader in, PrintWriter out) throws IOException {
+    in.close();
+    out.close();
```
Comment (esmevane) [inline src/main/java/server/Server.java]: Like I mentioned in another comment, the handling of your input and output might need to be distinguished.  Closing out your connections might also need to be a higher level concern (as an example, reading and writing could be handled by child objects, but closing connections happens after those objects have been invoked).

File: src/main/java/Response/Response.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package Response;
+
+import java.io.*;
+
+public class Response {
+  public String res200(String content) {
+    String headers = "HTTP/1.1 200 OK\r\n" +
+      "Content-Length: " + (content + "\r\n\r\n").length() + "\r\n" +
```
Comment (esmevane) [inline src/main/java/Response/Response.java]: I see that the `CRLF` is present as strings in a lot of your Response / Request object code.  Maybe that's a good opportunity for a constant on a top-level object?

File: src/main/java/server/Server.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  public void go() throws IOException {
+    serverSocket = createSocket();
+
+    while (true) {
+      clientSocket = clientSocket(serverSocket);
+
+      out = createWriter(clientSocket);
+      in = createReader(clientSocket);
+      handleRequest(in, out);
+      closeConnection(in, out);
```
Comment (esmevane) [inline src/main/java/server/Server.java]: I see in other parts of this file that you're invoking _both_ input and output in `handleRequest` and `closeConnection`.  I've made some suggestions to decouple those a little, but just to illustrate what I'm thinking of, the shape might be something like this:

```java
Socket clientSocket = serverSocket.accept();
BufferedReader input = createReader(clientSocket);
PrintWriter output = createWriter(clientSocket);

// This would turn the request into a `RequestContent` Java object.
//
RequestContent request = readRequest(input);

// This accepts a RequestContent object, and creates a custom ResponseContent object
//
ResponseContent response = new ResponseContent(request);

// This writes a response to the output
//
output.write(response.toString());

clientSocket.close();
```
